### ISSUE TITLE
[Turbo] Web Debug Toolbar script is blocked by CSP when navigating

### ISSUE THREAD

**Issue by DavidPetrasek** [symfony/ux#2134]
On initial page load the toolbar works fine, but is missing after visiting a link via Turbo Drive.

This is the console output from Google Chrome:
```
turbo.es2017-esm.js:4651 Refused to execute inline script because it violates the following Content Security Policy directive: "script-src https://127.0.0.1:5173 'nonce-ZkGUwME8QO5iHCXt1kgvHw==' 'unsafe-inline' 'nonce-e276b18f073898209eb899b46ffa3f01'". Note that 'unsafe-inline' is ignored if either a hash or nonce value is present in the source list.

renderElement @ turbo.es2017-esm.js:4651
assignNewBody @ turbo.es2017-esm.js:4826
(anonymní) @ turbo.es2017-esm.js:4732
preservingPermanentElements @ turbo.es2017-esm.js:1717
preservingPermanentElements @ turbo.es2017-esm.js:1814
replaceBody @ turbo.es2017-esm.js:4730
render @ turbo.es2017-esm.js:4682
renderSnapshot @ turbo.es2017-esm.js:1523
render @ turbo.es2017-esm.js:1483
```

The CSP is handled by the NelmioSecurityBundle. Here is an excerpt from its config file (nelmio_security.yaml):
```
...
csp:
        hosts: []
        content_types: []
        enforce:
            level1_fallback: false
            browser_adaptive:
                enabled: false
            # report-uri: '%router.request_context.base_url%/nelmio/csp/report'
            default-src:
                - 'none'
            style-src:
                - 'self'
                - https://www.gstatic.com https://fonts.googleapis.com                                # Google Fonts
            font-src:
                - 'self'
                - https://fonts.googleapis.com https://fonts.gstatic.com                              # Google Fonts
            img-src:
                - 'self' 
                - 'data:'
                - https://tile.openstreetmap.org/
            form-action:
                - 'self'
            frame-src:
                - 'self'
                - 'blob:'
                - https://www.google.com
            frame-ancestors:
                - 'self'
            object-src:
                - 'self'
            media-src:
                - 'self'
            connect-src:
                - 'self' 
                - 'wss:'
                - https://tile.openstreetmap.org/
            block-all-mixed-content: true # defaults to false, blocks HTTP content over HTTPS transport
            # upgrade-insecure-requests: true # defaults to false, upgrades HTTP requests to HTTPS transport
...
```

**Comment by smnandre** [comment]
Could you whitelist the WDT in dev ? As it wont be a problem/question in prod anyway ? 

You may need to look at the web profiler bundle, and maybe update something here? 
https://symfony.com/doc/current/profiler.html#updating-the-web-debug-toolbar-after-ajax-requests 

Or do you have something precise in mind in the TurboBundle, as i'm not seeing right now what could be done here ?

**Comment by DavidPetrasek** [comment]
I set the header according to the documentation. However, I still get the CSP error, and now I'm also seeing a new error:
`Uncaught TypeError: ajaxToolbarPanel is null`
![Snímek obrazovky_20240909_074026](https://github.com/user-attachments/assets/f86fb0af-3c71-4d02-89fa-4908dab1feab)

How can I whitelist WDT in NelmioSecurityBundle? Using a hash will only work temporarily until a new version is released.

**Comment by smnandre** [comment]
I think you may find better support on the NelmioSecurityBundle 

Not that i don't want to help here, but they are much more qualified on the CSP topic than me... and as the bug does not exist when the NelmioSecurityBundle is not used... i'm not sure how we can handle things here.. 

But if you have some ideas / suggestions on how to implement things on our side, please do open a PR (or an issue with clear explanation on what should be done here)

**Comment by DavidPetrasek** [comment]
Ok, I created a new issue here: https://github.com/symfony/symfony/issues/58223

**Comment by SimonChabrier** [comment]
Hello, this is indeed a big, very annoying problem with UxTurbo and the debugBar if we make strong CSP.

I will share with you my solution which is radical but works very well. We consider of course that if we have the debug bar it is that we are in development and that we never activate the debug bar in production.

Have a look in vendor, there is `'Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler'` and the class have a `public function disableCsp()`

So the solution is : 

## 1 Make a Service 

`<?php

namespace App\Service;

use Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler;

class DesableDebubBarCsp
{

    public function __construct(
        private ContentSecurityPolicyHandler $contentSecurityPolicyHandler,
    ) {}

    /**
     * Hoverhide ContentSecurityPolicyHandler from bundle symfony/web-profiler-bundle
     * 
     *
     * @return void
     */
    public function disableCsp(): void
    {
        $this->contentSecurityPolicyHandler->disableCsp();
    }
}`

## 2 Add an alias in services.yaml 

`Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler: '@web_profiler.csp.handler'` 

## 3 Make the Subscriber to inject the debug bar 

`use Symfony\Component\HttpKernel\KernelInterface;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;


class AjaxDebugBarSubscriber implements EventSubscriberInterface
{

    public function __construct(
        private KernelInterface $kernel,
        private DesableDebubBarCsp $csp
    ) {}

    public function onKernelResponse(ResponseEvent $event): void
    {
        // this remove the nonce generation
        $this->csp->disableCsp();

        if (!$this->kernel->isDebug()) {
            return;
        }

        $request = $event->getRequest();
        if (!$request->isXmlHttpRequest()) {
            return;
        }

        $response = $event->getResponse();

        $response->headers->set('Symfony-Debug-Toolbar-Replace', '1');
    }

    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::RESPONSE => 'onKernelResponse',
        ];
    }
}`

Finally, as the debug bar is injected after Nelmio Security has evaluated the state of the DOM (I thinks this), there are no more errors in the console after Ajax responses and we can continue to NOT ADD 'unsafe-inline' in the Nelmio Security configuration and continue to evaluate other possible problems related to other things without being bothered by the debug bar.

I tried differently before, someone had shared a small Javascript solution but it does not work systematically according to the rendering times and the problems persist. Here this solution is radical and for my part has completely solved the problem. I hope it will help you.

Do not forget that this is a voluntarily opened flaw and that the debugbar must never be activated in production for any reason.

That said, it is a real problem and it would be great if the symfony teams soon proposed a real solution that allows the injection of the debug bar in ajax responses to communicate with Turbo so that it is automated and does not force us to do radial processing like that.

You will tell me if it worked for you if someone decides to adopt this approach.
